commons/src/main/java/org/calrissian/accumulorecipes/commons/support/qfd/QfdHelper.java

```diff
@@ -196,63 +185,55 @@ public void save(Iterable<? extends T> items) {
-                                columnVisibility = new ColumnVisibility(visCache.getVis());
-                                colVisCache.put(visCache.getVis(), columnVisibility);
-                            }
-
-                            Collection<Map<String, Object>> metas = tupleByObjectCache.get(visCache);
-                            metas = Collections2.transform(metas, removeVisFunction);
-
-                            value.set(metas.size() > 0 ? metadataSerDe.serialize(metas) : EMPTY_VALUE.get());
-
-                            String id = buildId(item);
-                            forwardCF.set(id);
-                            forwardCQ.set(key + NULL_BYTE + aliasValue);
-                            invertedCF.set(PREFIX_FI + NULL_BYTE + key);
-                            invertedCQ.set(aliasValue + NULL_BYTE + id);
-
-                            // forward mutation
-                            shardMutation.put(forwardCF,
-                                forwardCQ,
-                                columnVisibility,
-                                value);
-
-                            // reverse mutation
-                            shardMutation.put(invertedCF,
-                                invertedCQ,
-                                columnVisibility,
-                                value);  // forward mutation
-                        }
+                    for (Tuple tuple : item.getTuples()) {
+                        String visibility = getVisibility(tuple.getMetadata(), "");
+                        String aliasValue = typeRegistry.getAlias(tuple.getValue()) + ONE_BYTE +
+                            typeRegistry.encode(tuple.getValue());
+
+                        ColumnVisibility columnVisibility = new ColumnVisibility(visibility);
+
+                        Map<String,Object> meta = tuple.getMetadata();
+
+                        value.set(metadataSerDe.serialize(meta));
+
+                        forwardCF.set(id);
+                        forwardCQ.set(tuple.getKey() + NULL_BYTE + aliasValue);
+                        invertedCF.set(PREFIX_FI + NULL_BYTE + tuple.getKey());
+                        invertedCQ.set(aliasValue + NULL_BYTE + id);
+
+
+                        Key key = new Key(new Text(shardId), forwardCF, forwardCQ, columnVisibility, 0);
+                        Value valuePart = new Value(value);
+                        visToKeyCache.put(columnVisibility, key);
+                        visToValueCache.put(columnVisibility, valuePart);
+
+                        shardMutation.put(invertedCF,
+                            invertedCQ,
+                            columnVisibility,
+                            value);
                     }
-                    shardWriter.addMutation(shardMutation);
+
+                  for(ColumnVisibility colVis : visToKeyCache.keySet()) {
+                    Value row = encodeRow(visToKeyCache.get(colVis), visToValueCache.get(colVis));
```

Comment:
My only concern with this method is the use to two data structures that need to be kept perfectly inline with each other (i.e. order matters) for the encodeRow to work.  At first this was a little confusing to me.

Instead I suggest using a Multimap<ColumnVisibility, Map.Entry<Key, Value>> to hold them both together.  This makes your intent more clear. For example,

Multimap<ColumnVisibility, Map.Entry<Key, Value>> visCache = ArrayListMultimap.create();
...
visCache.put(columnVisibility, Maps.immutableEntry(key, valuePart));
...
encodeRow(visCache.get(colVis))

Additionally, now incode row can just take in one iterable (or List) which also maps the key and value together.